### PR TITLE
implemented cell identity type for later use

### PR DESCRIPTION
**Pull request by wowtor**
Type imported from other repo; will be useful later on

### CODE REVIEW

File: telcell/cell_identity.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+    @staticmethod
+    def create(
+        radio: Optional[str] = None,
+        mcc: Optional[int] = None,
+        mnc: Optional[int] = None,
+        lac: Optional[int] = None,
+        ci: Optional[int] = None,
+        eci: Optional[int] = None,
+    ) -> "CellIdentity":
```
Comment (akaIDIOT) [inline telcell/cell_identity.py:31]: Considering they're all optional and the test code passes all of them by name, would it make sense to force them to be keyword arguments?
```suggestion
    def create(*,
        radio: Optional[str] = None,
        mcc: Optional[int] = None,
        mnc: Optional[int] = None,
        lac: Optional[int] = None,
        ci: Optional[int] = None,
        eci: Optional[int] = None,
    ) -> CellIdentity:
```

Also, typing expression are evaluated later, the return type can be unquoted.

File: telcell/cell_identity.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+            * [https://arimas.com/cgi-ecgi/]
+            * [http://www.techtrained.com/the-ultimate-cheat-sheet-for-lte-identifiers/]
+
+        Parameters:
+            mcc: Mobile Country Code (MCC, 3 digits)
+            mnc: Mobile Network Code (MNC, 2 digits)
+        """
+        assert mcc is None or isinstance(mcc, int), "mcc must be of type `int`"
+        assert mnc is None or isinstance(mnc, int), "mnc must be of type `int`"
```
Comment (akaIDIOT) [inline telcell/cell_identity.py:82]: Other bits use an `if` statement. Considering that assertions outside of tests are frowned upon, maybe use an `if` here too?

File: telcell/cell_identity.py
Comment (akaIDIOT) [inline telcell/cell_identity.py:1]: It feels like this file might be easier to follow if the different identities were dataclasses (for things like `__eq__` and the `radio` attribute, for example), but I'm not 100% all of the same things would still hold (though tests should be able to capture that :innocent:).

File: telcell/cell_identity.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
+import re
+from typing import Optional, Any
+
+RADIO_GSM = "GSM"
+RADIO_UMTS = "UMTS"
+RADIO_LTE = "LTE"
+RADIO_NR = "NR"
```
Comment (akaIDIOT) [inline telcell/cell_identity.py:7]: Potential candidate for an enum maybe, even though it wouldn't change too much.
```suggestion
class Radio(Enum):
    GSM = "GSM"
    UMTS = "UMTS"
    LTE = "LTE"
    NR = "NR"
```

File: tests/test_cell_identity.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    ("204-1-2-2000", CellIdentity.create(mcc=204, mnc=1, lac=2, ci=2000)),
+    ("GSM/204-1-2-2000", CellIdentity.create(radio=RADIO_GSM, mcc=204, mnc=1, lac=2, ci=2000)),
+    ("UMTS/204-1-2-2000", CellIdentity.create(radio=RADIO_UMTS, mcc=204, mnc=1, lac=2, ci=2000)),
+    ("LTE/204-1-2000", CellIdentity.create(radio=RADIO_LTE, mcc=204, mnc=1, eci=2000)),
+    ("NR/204-1-2000", CellIdentity.create(radio=RADIO_NR, mcc=204, mnc=1, eci=2000)),
+    ("GSM/204-1-?-?", CellIdentity.create(radio=RADIO_GSM, mcc=204, mnc=1)),
+]
+
+
+def test_operators(testdata_path):
```
Comment (akaIDIOT) [inline tests/test_cell_identity.py:19]: Test does not use `testdata_path` fixture.

File: tests/test_cell_identity.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    ("204-1", CellIdentity.create(mcc=204, mnc=1)),
+    ("GSM/204-1", CellIdentity.create(radio=RADIO_GSM, mcc=204, mnc=1)),
+    ("204-2", CellIdentity.create(mcc=204, mnc=2)),
+    ("204-1-2000", CellIdentity.create(mcc=204, mnc=1, eci=2000)),
+    ("204-1-2-2000", CellIdentity.create(mcc=204, mnc=1, lac=2, ci=2000)),
+    ("GSM/204-1-2-2000", CellIdentity.create(radio=RADIO_GSM, mcc=204, mnc=1, lac=2, ci=2000)),
+    ("UMTS/204-1-2-2000", CellIdentity.create(radio=RADIO_UMTS, mcc=204, mnc=1, lac=2, ci=2000)),
+    ("LTE/204-1-2000", CellIdentity.create(radio=RADIO_LTE, mcc=204, mnc=1, eci=2000)),
+    ("NR/204-1-2000", CellIdentity.create(radio=RADIO_NR, mcc=204, mnc=1, eci=2000)),
+    ("GSM/204-1-?-?", CellIdentity.create(radio=RADIO_GSM, mcc=204, mnc=1)),
```
Comment (akaIDIOT) [inline tests/test_cell_identity.py:15]: Would the first, second and last entries effectively encode the same cellid? If so: would that be worth testing for?

**Review by akaIDIOT — APPROVED** [review]

File: telcell/cell_identity.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        elif ci is not None and lac is not None and eci is None:
+            # we have `lac` and `ci` but not `eci` --> guess GSM/UMTS
+            return CellGlobalIdentity(mcc, mnc, lac, ci)
+        elif eci is None and lac is None and ci is None:
+            return CellIdentity(mcc, mnc)  # guess it's a cell
+        else:
+            raise ValueError("either `ci` or `eci` should be provided, but not both")
+
+    @staticmethod
+    def parse(spec: str) -> "CellIdentity":
```
[the remark this answers is not in the document]
Comment (wowtor) [reply telcell/cell_identity.py:56]: ok

File: telcell/cell_identity.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        return hash(str(self))
+
+
+class CellGlobalIdentity(CellIdentity):
+    """
+    The LAC is 16 bits (or 5 digits). The CI in GSM and CDMA/UMTS networks is
+    16 bits (or 5 digits).
+    """
+
+    __hash__ = CellIdentity.__hash__
```
[the remark this answers is not in the document]
Comment (wowtor) [reply telcell/cell_identity.py:131]: See here: https://stackoverflow.com/questions/53518981/inheritance-hash-sets-to-none-in-a-subclass

File: telcell/cell_identity.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+
+class NRCell(EutranCellGlobalIdentity):
+    __hash__ = EutranCellGlobalIdentity.__hash__
+
+    @property
+    def radio(self) -> str:
+        return RADIO_NR
+
+    def _asdict(self) -> dict:
```
[the remark this answers is not in the document]
Comment (wowtor) [reply telcell/cell_identity.py:291]: removed

File: telcell/cell_identity.py (hunk 2 above)
Comment (wowtor) [reply telcell/cell_identity.py:90]: done, although so far nobody convinced me that using `assert` is bad practice

**Pull request merged**